### PR TITLE
fix system.accessviolationexception

### PR DESCRIPTION
**Pull request by ispysoftware**
Some header files are causing a System.AccessViolationException when parsing. This fixes that.

Example header that caused the issue:
[test.zip](https://github.com/mono/CppSharp/files/1459467/test.zip)

### CODE REVIEW

File: src/CppParser/Parser.cpp [hunk 1]
```diff
@@ -887,9 +887,11 @@ static bool HasLayout(const clang::RecordDecl* Record)
 
 bool Parser::IsSupported(const clang::NamedDecl* ND)
 {
-    return !c->getSourceManager().isInSystemHeader(ND->getLocStart()) ||
-        (llvm::isa<clang::RecordDecl>(ND) &&
-         supportedStdTypes.find(ND->getName()) != supportedStdTypes.end());
+
+	return !ND->getSourceRange().isValid() ||
+		!c->getSourceManager().isInSystemHeader(ND->getLocStart()) ||
```
Comment (ddobrev) [inline src/CppParser/Parser.cpp]: Please keep the alignment we had before - just 4 spaces per line. Basically, the code should be aligned the same way as before, with the exception of the newly added first line.

**Comment by ddobrev** [comment]
Thank you for your code. I would like to ask you to make some additional changes besides the review itself:
1. Please add your test to your commit and ensure it crashes with and works without, your fix;
2. Please combine all commits into a single one. In case you use TortoiseGit, it's really easy to do from the log. Then push force.

**Comment by ispysoftware** [comment]
i'll try to figure out why it's crashing and recommit if i find it

**Comment by ddobrev** [comment]
In addition, please change your commit message to describe exactly what you've fixed.

**Comment by ispysoftware** [comment]
I don't know how to combine them into one commit - there are so many commits because the code you sent me had missing brackets and visual studio kept auto-formatting the file.

**Comment by ddobrev** [comment]
@ispysoftware I understand. Please just send your test and I'll fix it myself. Your test reproduces both crashes, correct?

**Comment by ispysoftware** [comment]
i've uploaded the test to the first comment. It only reproduces the first crash. I have no idea what caused the second one but the check to ensure the type that you added fixed it.

**Comment by ddobrev** [comment]
I understand. Could you please try reproducing the second crash too? The same way as before - you start with the crashing header, use dumpColor to find the declaration and then start reducing it until it stops.

**Comment by ispysoftware** [comment]
this is the dumpcolor output when the ND is nullptr. Maybe this will mean something to you?
[compare.zip](https://github.com/mono/CppSharp/files/1459541/compare.zip)


LinkageSpecDecl 0x1866b3c0 <C:\webrtc\webrtc-checkout\src\third_party\libyuv\include/libyuv/compare.h:18:1, line:102:1> line:18:8 C
|-FunctionDecl 0x18672650 <line:23:1, col:60> col:8 HashDjb2 'uint32 (const uint8 *, uint64, uint32)'
| |-ParmVarDecl 0x1866b488 <col:17, col:30> col:30 src 'const uint8 *'
| |-ParmVarDecl 0x1866b4e0 <col:35, col:42> col:42 count 'uint64':'unsigned long long'
| `-ParmVarDecl 0x1866b520 <col:49, col:56> col:56 seed 'uint32':'unsigned int'
|-FunctionDecl 0x18672820 <line:27:1, line:29:40> line:27:8 ComputeHammingDistance 'uint64 (const uint8 *, const uint8 *, int)'
| |-ParmVarDecl 0x186726d8 <col:31, col:44> col:44 src_a 'const uint8 *'
| |-ParmVarDecl 0x18672720 <line:28:31, col:44> col:44 src_b 'const uint8 *'
| `-ParmVarDecl 0x18672768 <line:29:31, col:35> col:35 count 'int'
|-FunctionDecl 0x18672a48 <line:34:1, col:76> col:8 ARGBDetect 'uint32 (const uint8 *, int, int, int)'
| |-ParmVarDecl 0x186728a8 <col:19, col:32> col:32 argb 'const uint8 *'
| |-ParmVarDecl 0x186728f0 <col:38, col:42> col:42 stride_argb 'int'
| |-ParmVarDecl 0x18672938 <col:55, col:59> col:59 width 'int'
| `-ParmVarDecl 0x18672980 <col:66, col:70> col:70 height 'int'
|-FunctionDecl 0x18672bc0 <line:38:1, col:79> col:8 ComputeSumSquareError 'uint64 (const uint8 *, const uint8 *, int)'
| |-ParmVarDecl 0x18672ad0 <col:30, col:43> col:43 src_a 'const uint8 *'
| |-ParmVarDecl 0x18672b18 <col:50, col:63> col:63 src_b 'const uint8 *'
| `-ParmVarDecl 0x18672b60 <col:70, col:74> col:74 count 'int'
|-FunctionDecl 0x18672e98 <line:41:1, line:46:45> line:41:8 ComputeSumSquareErrorPlane 'uint64 (const uint8 *, int, const uint8 *, int, int, int)'
| |-ParmVarDecl 0x18672c48 <col:35, col:48> col:48 src_a 'const uint8 *'
| |-ParmVarDecl 0x18672c90 <line:42:35, col:39> col:39 stride_a 'int'
| |-ParmVarDecl 0x18672cd8 <line:43:35, col:48> col:48 src_b 'const uint8 *'
| |-ParmVarDecl 0x18672d20 <line:44:35, col:39> col:39 stride_b 'int'
| |-ParmVarDecl 0x18672d68 <line:45:35, col:39> col:39 width 'int'
| `-ParmVarDecl 0x18672db0 <line:46:35, col:39> col:39 height 'int'
|-VarDecl 0x18672f28 <line:48:1, col:29> col:18 kMaxPsnr 'const int' static cinit
| `-IntegerLiteral 0x18672f60 <col:29> 'int' 128
|-FunctionDecl 0x186730c0 <line:51:1, col:53> col:8 SumSquareErrorToPsnr 'double (uint64, uint64)'
| |-ParmVarDecl 0x18672fc8 <col:29, col:36> col:36 sse 'uint64':'unsigned long long'
| `-ParmVarDecl 0x18673008 <col:41, col:48> col:48 count 'uint64':'unsigned long long'
|-FunctionDecl 0x18673390 <line:54:1, line:59:32> line:54:8 CalcFramePsnr 'double (const uint8 *, int, const uint8 *, int, int, int)'
| |-ParmVarDecl 0x18673140 <col:22, col:35> col:35 src_a 'const uint8 *'
| |-ParmVarDecl 0x18673188 <line:55:22, col:26> col:26 stride_a 'int'
| |-ParmVarDecl 0x186731d0 <line:56:22, col:35> col:35 src_b 'const uint8 *'
| |-ParmVarDecl 0x18673218 <line:57:22, col:26> col:26 stride_b 'int'
| |-ParmVarDecl 0x18673260 <line:58:22, col:26> col:26 width 'int'
| `-ParmVarDecl 0x186732a8 <line:59:22, col:26> col:26 height 'int'
|-FunctionDecl 0x18677940 <line:62:1, line:75:27> line:62:8 I420Psnr 'double (const uint8 *, int, const uint8 *, int, const uint8 *, int, const uint8 *, int, const uint8 *, int, const uint8 *, int, int, int)'
| |-ParmVarDecl 0x18673420 <col:17, col:30> col:30 src_y_a 'const uint8 *'
| |-ParmVarDecl 0x18673468 <line:63:17, col:21> col:21 stride_y_a 'int'
| |-ParmVarDecl 0x186734b0 <line:64:17, col:30> col:30 src_u_a 'const uint8 *'
| |-ParmVarDecl 0x186734f8 <line:65:17, col:21> col:21 stride_u_a 'int'
| |-ParmVarDecl 0x18673540 <line:66:17, col:30> col:30 src_v_a 'const uint8 *'
| |-ParmVarDecl 0x18673588 <line:67:17, col:21> col:21 stride_v_a 'int'
| |-ParmVarDecl 0x186735d0 <line:68:17, col:30> col:30 src_y_b 'const uint8 *'
| |-ParmVarDecl 0x18673618 <line:69:17, col:21> col:21 stride_y_b 'int'
| |-ParmVarDecl 0x18677688 <line:70:17, col:30> col:30 src_u_b 'const uint8 *'
| |-ParmVarDecl 0x186776d0 <line:71:17, col:21> col:21 stride_u_b 'int'
| |-ParmVarDecl 0x18677718 <line:72:17, col:30> col:30 src_v_b 'const uint8 *'
| |-ParmVarDecl 0x18677760 <line:73:17, col:21> col:21 stride_v_b 'int'
| |-ParmVarDecl 0x186777a8 <line:74:17, col:21> col:21 width 'int'
| `-ParmVarDecl 0x186777f0 <line:75:17, col:21> col:21 height 'int'
|-FunctionDecl 0x18677bc8 <line:78:1, line:83:32> line:78:8 CalcFrameSsim 'double (const uint8 *, int, const uint8 *, int, int, int)'
| |-ParmVarDecl 0x186779f0 <col:22, col:35> col:35 src_a 'const uint8 *'
| |-ParmVarDecl 0x18677a38 <line:79:22, col:26> col:26 stride_a 'int'
| |-ParmVarDecl 0x18677a80 <line:80:22, col:35> col:35 src_b 'const uint8 *'
| |-ParmVarDecl 0x18677ac8 <line:81:22, col:26> col:26 stride_b 'int'
| |-ParmVarDecl 0x18677b10 <line:82:22, col:26> col:26 width 'int'
| `-ParmVarDecl 0x18677b58 <line:83:22, col:26> col:26 height 'int'
`-FunctionDecl 0x18678090 <line:86:1, line:99:27> line:86:8 I420Ssim 'double (const uint8 *, int, const uint8 *, int, const uint8 *, int, const uint8 *, int, const uint8 *, int, const uint8 *, int, int, int)'
  |-ParmVarDecl 0x18677c58 <col:17, col:30> col:30 src_y_a 'const uint8 *'
  |-ParmVarDecl 0x18677ca0 <line:87:17, col:21> col:21 stride_y_a 'int'
  |-ParmVarDecl 0x18677ce8 <line:88:17, col:30> col:30 src_u_a 'const uint8 *'
  |-ParmVarDecl 0x18677d30 <line:89:17, col:21> col:21 stride_u_a 'int'
  |-ParmVarDecl 0x18677d78 <line:90:17, col:30> col:30 src_v_a 'const uint8 *'
  |-ParmVarDecl 0x18677dc0 <line:91:17, col:21> col:21 stride_v_a 'int'
  |-ParmVarDecl 0x18677e08 <line:92:17, col:30> col:30 src_y_b 'const uint8 *'
  |-ParmVarDecl 0x18677e50 <line:93:17, col:21> col:21 stride_y_b 'int'
  |-ParmVarDecl 0x18677e98 <line:94:17, col:30> col:30 src_u_b 'const uint8 *'
  |-ParmVarDecl 0x18677ee0 <line:95:17, col:21> col:21 stride_u_b 'int'
  |-ParmVarDecl 0x18677f28 <line:96:17, col:30> col:30 src_v_b 'const uint8 *'
  |-ParmVarDecl 0x18677f70 <line:97:17, col:21> col:21 stride_v_b 'int'
  |-ParmVarDecl 0x18677fb8 <line:98:17, col:21> col:21 width 'int'
  `-ParmVarDecl 0x18678000 <line:99:17, col:21> col:21 height 'int'

**Comment by ddobrev** [comment]
I cannot tell right now but that's not important. What is is that you've found the offending header. Just cut it down until you have a small reproduction, just like the previous time.
It's really late here and I have to go. But just leave me the second test and I'll fix both crashes in the morning. Good night to you.

**Comment by ispysoftware** [comment]
ok here you go
[test2.zip](https://github.com/mono/CppSharp/files/1459559/test2.zip)

**Comment by ddobrev** [comment]
Fixed by https://github.com/mono/CppSharp/commit/9c2f62c3ebee47f161a75a3a8e2616fa8700e37f .

**Pull request closed without merging**